### PR TITLE
Cnn wet dry

### PR DESCRIPTION
**Pull request by jpolz**
Add a wet/dry function using the cnn method.

- [x] Add trained cnn files to subdirectory of wet/dry

- [x] Add function for (txrx_ch_1,txrx_ch_2)-->wet/dry

- [x] Add test for wet dry classification

Reference: https://doi.org/10.5194/amt-13-3835-2020

### CODE REVIEW

**Comment by jpolz** [comment]
Test is done, but not perfect. Problem is that a new version of the cnn would change the results, therefore I just check if the output length of the array is fine.

**Review by cchwala — CHANGES_REQUESTED** [review]

**Comment by cchwala** [comment]
> Test is done, but not perfect. Problem is that a new version of the cnn would change the results, therefore I just check if the output length of the array is fine.

This does not seem to be very robust against unexpected changes. If a new version of the CNN gets added, then the test has to be adapted, or tests for individual "versions" of the CNN might be an option.

File: pycomlink/processing/wet_dry/cnn.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    df = pd.DataFrame()
+    df["trsl1"] = trsl_channel_1
+    df["trsl2"] = trsl_channel_2
+    idx = df.index
+    df["med1"] = df["trsl1"].rolling(72 * 60, min_periods=2 * 60, center=False).median()
+    df["med2"] = df["trsl2"].rolling(72 * 60, min_periods=2 * 60, center=False).median()
+    df["trsl1"] = df["trsl1"].sub(df["med1"])
+    df["trsl2"] = df["trsl2"].sub(df["med2"])
+    df = df.dropna()
```
[the remark this answers is not in the document]
Comment (jpolz) [reply pycomlink/processing/wet_dry/cnn.py:84]: NaNs can not be handled by the CNN, therefore they are replaced by -9999. afterwards -9999 will again lead to a NaN in the prediction. It is not trivial to do predictions just for the samples containing no NaNs.

File: pycomlink/processing/wet_dry/cnn.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            or -9999 in df["trsl2"].values[i - 151 : i + 31]
+        ):
+            cnn_pred[i] = np.nan
+
+    df["prediction"] = np.concatenate(
+        (np.repeat(np.nan, 150), cnn_pred, np.repeat(np.nan, 29)), axis=0
+    )
+
+    df = df.reindex(idx, fill_value=np.nan)
+    return df.prediction.values > threshold
```
[the remark this answers is not in the document]
Comment (jpolz) [reply pycomlink/processing/wet_dry/cnn.py:115]: Should be easy! But it has to be more obvious to not cause troubles later when the function is used blindly with other functions. Maybe `return_raw_predictions=True`

File: pycomlink/processing/wet_dry/cnn.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    References
+    ----------
+    .. [1] Polz, J., Chwala, C., Graf, M., and Kunstmann, H.: Rain event detection in commercial microwave link
+    attenuation data using convolutional neural networks, Atmos. Meas. Tech., 13, 3835–3853,
+    https://doi.org/10.5194/amt-13-3835-2020, 2020.
+    """
+
+    # load json and create model
+    json_file = open(modeljson_fn, "r")
```
[the remark this answers is not in the document]
Comment (jpolz) [reply pycomlink/processing/wet_dry/cnn.py:61]: it makes sense to do this anyways since the model data is small.

**Comment by jpolz** [comment]
> > Test is done, but not perfect. Problem is that a new version of the cnn would change the results, therefore I just check if the output length of the array is fine.
> 
> This does not seem to be very robust against unexpected changes. If a new version of the CNN gets added, then the test has to be adapted, or tests for individual "versions" of the CNN might be an option.

What could be a meaningful test then? It is also risky to update the exact expected output for every new model version that we implement.

File: pycomlink/processing/wet_dry/cnn.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpolz) [reply pycomlink/processing/wet_dry/cnn.py:61]: 650 kB in total. I move it to outside the function.

File: pycomlink/processing/wet_dry/cnn.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            ]
+        ),
+        0,
+        -1,
+    )
+    cnn_pred = np.ravel(model.predict(x_fts, batch_size=batch_size, verbose=verbose))
+
+    for i in range(len(cnn_pred)):
+        if (
+            -9999 in df["trsl1"].values[i - 151 : i + 31]
```
[the remark this answers is not in the document]
Comment (jpolz) [reply pycomlink/processing/wet_dry/cnn.py]: Done.

**Comment by jpolz** [comment]
> I would test using a given example trsl time series and a "known" expected result which you have to get from running your code and checking yourself if the result makes sense. There is no way (that I know of...) to write a generic test for this sort of data processing functions. The correctness, and this is what tests are about, can only be checked by applying it to an example dataset.
> 
> The problem is that e.g. updates to your imported dependencies might change your results. You want to make sure that this is caught by a test. For a new model, you will have to update the test. Yes. But I do not see any problem with this. You commit some code for the new model, anyway, so you can also update the test then.

This should be resolved now. Although I do not test on a real TRSL time series the test should test if the model always performs the exact same numerical operations.

File: pycomlink/processing/wet_dry/cnn.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpolz) [reply pycomlink/processing/wet_dry/cnn.py]: I replaced this routine by `df.fillna()`

**Comment by cchwala** [comment]
Tests do not run because your new dependencies are not in https://github.com/pycomlink/pycomlink/blob/v0.3alpha/.travis.yml
Just add them there, or even better, parser the dependencies from `requirements.txt`.

**Comment by jpolz** [comment]
Done

**Review by cchwala — APPROVED** [review]

**Comment by cchwala** [comment]
Looks good!

CNN wet-dry classification in pycomlink! 🚀

**Pull request merged**